lib/galaxy/workflow/modules.py

```diff
@@ -626,77 +723,316 @@ class InputDataCollectionModule(InputModule):
+
+            when_true = ConditionalWhen()
+            when_true.value = "true"
+            when_true.inputs = OrderedDict()
+            when_true.inputs["default"] = specify_default_cond
+
+            when_false = ConditionalWhen()
+            when_false.value = "false"
+            when_false.inputs = OrderedDict()
+
+            optional_cases = [when_true, when_false]
+            optional_cond.cases = optional_cases
+
+            if param_type == "text":
+                restrict_how_source = dict(name="how", label="Restrict Text Values?", type="select")
+                if parameter_def.get("restrictions") is not None:
+                    restrict_how_value = "staticRestrictions"
+                elif parameter_def.get("restrictOnConnections") is True:
+                    restrict_how_value = "onConnections"
+                elif parameter_def.get("suggestions") is not None:
+                    restrict_how_value = "staticSuggestions"
+                else:
+                    restrict_how_value = "none"
+                log.info("parameter_def [%s], how [%s]" % (parameter_def, restrict_how_value))
+                restrict_how_source["options"] = [
+                    {"value": "none", "label": "Do not specify restrictions (default).", "selected": restrict_how_value == "none"},
+                    {"value": "onConnections", "label": "Attempt restriction based on connections.", "selected": restrict_how_value == "onConnections"},
+                    {"value": "staticRestrictions", "label": "Provide list of all possible values.", "selected": restrict_how_value == "staticRestrictions"},
+                    {"value": "staticSuggestions", "label": "Provide list of suggested values.", "selected": restrict_how_value == "staticSuggestions"},
+                ]
+                restrictions_cond = Conditional()
+                restrictions_how = SelectToolParameter(None, restrict_how_source)
+                restrictions_cond.name = "restrictions"
+                restrictions_cond.test_param = restrictions_how
+
+                when_restrict_none = ConditionalWhen()
+                when_restrict_none.value = "none"
+                when_restrict_none.inputs = OrderedDict()
+
+                when_restrict_connections = ConditionalWhen()
+                when_restrict_connections.value = "onConnections"
+                when_restrict_connections.inputs = OrderedDict()
+
+                when_restrict_static_restrictions = ConditionalWhen()
+                when_restrict_static_restrictions.value = "staticRestrictions"
+                when_restrict_static_restrictions.inputs = OrderedDict()
+
+                when_restrict_static_suggestions = ConditionalWhen()
+                when_restrict_static_suggestions.value = "staticSuggestions"
+                when_restrict_static_suggestions.inputs = OrderedDict()
+
+                # Repeats don't work - so use common separated list for now.
```

Comment:
Why do repeats not work?

Reply:
They don't - I don't know why exactly. I spent a few hours trying to walk through what was happening with the parameter expansion logic and trying a variety of tweaks and was pretty convinced it wasn't a trivial amount of code away from working.

We could do some cool stuff around allowing users to declare expressions with multiple inputs and outputs if we could get that to work.